azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
 
         else:
             super(UseSecurityClientStage, self)._run_op(op)
+
+
+class CommonProvisioningStage(PipelineStage):
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_error(self, provisioning_op, prov_op_name, error):
```

Comment:
>_get_error [](start = 8, length = 10)

This function concerns me because it's doing more than just getting the error.  It's also deciding to swallow all status codes 429 and above.  I understand why you might consider those to not be errors, but this is the wrong place to make that decision.

Reply:
is there a better place for this ? may be we need to talk regarding this. According to me , >= 429 are only things where we need to take a different decision.

---
In reply to: [364368562](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r364368562) [](ancestors = 364368562)

Reply:
It's just a weird flow.  This function, which, according to the name, returns an error from the status_code.  But, as a sneaky hidden side-effect at the very bottom, turns some errors into success.  Then, in PollingStatusStage._run_op, it uses the status_code to handle these special "error that was turned into success" cases.

So, _get_error can return one of two things:
1. "definitely an error" if it return an Error object.
2. "maybe success" if it returns None.  
Callers can never look at the return value from this function to know if the operation succeeded.  They also sometimes need to look at the HTTP status code and sometimes they need to look at the provisioning status.    This is confusing.  This function doesn't make things clearer. The caller still needs add code to handle "maybe success" by looking at the status code and provisioning status.  

It would cleaner if this function could return an answer that callers could rely on.  One possibility is to return one of three things:
1. Operation is still in progress.   We either got 429+ or we got status=="assigning".  Caller knows that it needs to poll again.
1. Definitely an error.  We got 300-428 or we got status=="failed"  Caller knows that it needs to fail the operation.  
2. Definitely success.  We got status=="assigned".  Caller knows that the operation is a success.  


---
In reply to: [365639409](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365639409) [](ancestors = 365639409,364368562)

Reply:
pure error happens when error was returned from the lower levels or i am getting a code in between 300 to 428. Anything after 429 would need to be retried.
hence i designed this accordingly.

if we need to take a decision based on the return value of this function then this function can not return a status code sometimes and a string status other times. I have to assign another status called "retrying" whenever i receive >= 429. For the registration stage these 2 things ("assigning" and "retrying") means 2 separate things.

and for 300-428 i have to assign something like "error" but we already have a valid status called "error". I don't know what to call this 300-428.

let me know if this sounds fine. 



---
In reply to: [366481124](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r366481124) [](ancestors = 366481124,365639409,364368562)